### ISSUE TITLE
Persistent store for Key Server [Drop 1]

### ISSUE THREAD

**Issue by LordQ1** [boltlabs-inc/key-mgmt#346]
**Overview:** We have discussed replacing our persistent storage on the Key Server with something more robust, while also allowing users of our framework to plug in their own persistent storage in the future. This Epic attempts to capture some of the high-level requirements. There will be sub-issues to track the individual requirements -- and, we expect that some of the associated sub-issues will span more than one of our release cycles.

- [x] Persistent store abstraction layer #338
- [x] Select open source persistent storage product for development #372
- [x] Integrate and test persistent storage product with Key Server #378 #435 
- [x] Define a better database schema for stored secrets #339
- [ ] Re-visit database errors #366

**Initial Thoughts on Requirements:**

- Provide a database abstraction layer between our software framework and the actual persistence store -- implementors should be able to swap out the persistence store by changing the abstraction layer, while not imposing effects upon the (higher layers of) software framework
- Select an initial persistence store (specific vendor product) for our reference implementation 
   - Our initial selection should be an open source implementation that can be tested as compliant with our software framework
   - Our selection should not impact others from plugging in a different implementation that might not be open source
- The selected product for our reference implementation should offer the following deployment features:
   - Distribution: Support for local deployments (on the same machine or VM as the Key Server) as well as remote deployments (separate from the Key Server)
   - Scalability: Support for handling a large number of transactions from multiple simultaneous threads and processes
   - Availability: Support for active-active and active-passive deployment models (i.e., multiple servers interfacing to the same persistent store for high availability models)
   - Backup/Recovery: Support for full and incremental backups
- While each of these deployment features should be supported by the selected product, we should not impose the full feature set on any deployment 
   - In other words, it is an exercise for the deployment team and/or security officer (not Bolt) to decide which of the deployment features they intend to support for their solution, which could influence their vendor selection
